tests/callbacks_test.cpp

```diff
@@ -0,0 +1,382 @@
+
+  const bool hasBeginEpoch =  HasBeginEpoch<BeginEpochTestFunction,
+      TypedForms<StandardSGD, BeginEpochTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  const bool hasEndEpoch =  HasEndEpoch<BeginEpochTestFunction,
+      TypedForms<StandardSGD, BeginEpochTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  REQUIRE(hasEvaluate == false);
+  REQUIRE(hasGradient == false);
+  REQUIRE(hasBeginEpoch == true);
+  REQUIRE(hasEndEpoch == false);
+}
+
+/**
+ * Make sure we don't add any functions if we only have EndEpoch().
+ */
+TEST_CASE("CallbacksEndEpochOnlyTest", "[CallbacksTest]")
+{
+  const bool hasEvaluate =  HasEvaluate<EndEpochTestFunction,
+      TypedForms<StandardSGD, EndEpochTestFunction,
+      arma::mat>::template EvaluateForm>::value;
+
+  const bool hasGradient =  HasGradient<EndEpochTestFunction,
+      TypedForms<StandardSGD, EndEpochTestFunction,
+      arma::mat, arma::mat>::template GradientForm>::value;
+
+  const bool hasBeginEpoch =  HasBeginEpoch<EndEpochTestFunction,
+      TypedForms<StandardSGD, EndEpochTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  const bool hasEndEpoch =  HasEndEpoch<EndEpochTestFunction,
+      TypedForms<StandardSGD, EndEpochTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  REQUIRE(hasEvaluate == false);
+  REQUIRE(hasGradient == false);
+  REQUIRE(hasBeginEpoch == false);
+  REQUIRE(hasEndEpoch == true);
+}
+
+/**
+ * Make sure we have all callbacks.
+ */
+TEST_CASE("CallbacksCompleteFunctionsTest", "[CallbacksTest]")
+{
+  const bool hasEvaluate =  HasEvaluate<CompleteTestFunction,
+      TypedForms<StandardSGD, CompleteTestFunction,
+      arma::mat>::template EvaluateForm>::value;
+
+  const bool hasGradient =  HasGradient<CompleteTestFunction,
+      TypedForms<StandardSGD, CompleteTestFunction,
+      arma::mat, arma::mat>::template GradientForm>::value;
+
+  const bool hasBeginEpoch =  HasBeginEpoch<CompleteTestFunction,
+      TypedForms<StandardSGD, CompleteTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  const bool hasEndEpoch =  HasEndEpoch<CompleteTestFunction,
+      TypedForms<StandardSGD, CompleteTestFunction,
+      arma::mat>::template BeginEpochForm>::value;
+
+  REQUIRE(hasEvaluate == true);
+  REQUIRE(hasGradient == true);
+  REQUIRE(hasBeginEpoch == true);
+  REQUIRE(hasEndEpoch == true);
+}
+
+/**
+ * Make sure the EarlyStopAtMinLoss callback will stop the optimization process.
+ */
+TEST_CASE("EarlyStopAtMinLossCallbackTest", "[CallbacksTest]")
+{
+  SGDTestFunction f;
+  arma::mat coordinates = f.GetInitialPoint();
+
+  // Instantiate the optimizer with a number of iterations that will take a
+  // long time to finish.
+  StandardSGD s(0.0003, 1, 10000000000, -10, true);
+
+  // The optimization process should return in one second.
+  const double result = s.Optimize(f, coordinates, EarlyStopAtMinLoss(100));
+
+  REQUIRE(s.Terminate());
```

Comment:
Ah, I guess if we remove `Terminate()` then this test gets harder.  I guess that we could use a really simple quadratic function to optimize, and start very far from the optimum with a very small step size.  Then, if termination happens very early, then we would expect the point to be far closer to the starting point than to the optimum.